### PR TITLE
[DOC] Self-SL & SupCon for semantic segmentation

### PR DESCRIPTION
**Pull request by sungchul1**
- [x] Self-SL
- [x] SupCon

### CODE REVIEW

**Comment by sungchul1** [comment]
## SupCon
image

## Self-SL
image

**Review by harimkang — DISMISSED** [review]

**Review by harimkang — APPROVED** [review]
LGTM!

**Review by sungmanc — APPROVED** [review]

**Comment by goodsong81** [comment]
@kprokofi @wonjuleee Could you leave you review comment even after merge so that @sungchul2 could refine this in another PR? Emily requested quick merge to address advices from Geti UX/doc team.

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst [hunk 1]
```diff
@@ -25,7 +25,58 @@ For the supervised training we use the following algorithms components:
 
 - ``Loss function``: We use standart `Cross Entropy Loss <https://en.wikipedia.org/wiki/Cross_entropy>`_  to train a model.
 
-- Additionally, we use the **early stopping** to add adaptability to the training pipeline and prevent overfitting.
+- ``Training technique``
```
Comment (kprokofi) [inline docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:28]: - ``Additional training techniques`` or ``Additional training tricks``

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst [hunk 2]
```diff
@@ -25,7 +25,58 @@ For the supervised training we use the following algorithms components:
...
+    - ``Early stopping``: To add adaptability to the training pipeline and prevent overfitting. You can use early stopping like the below command.
+
+      .. code-block::
+
+        $ otx train {TEMPLATE} ... \
+            params \
+            --learning_parameters.enable_early_stopping=True \      # is early stopping used
+            --learning_parameters.early_stop_start=3 \              # the number of epochs (iters) in which early stopping proceeds
```
Comment (kprokofi) [inline docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:36]: Please, use only ``--learning_parameters.enable_early_stopping=True``  for the example. Also, the early stopping is on by default. Maybe it is even redundant and misleading information. In the tutorials we have some examples how to modify parameters. I suggest deleting it at all

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst [hunk 3]
```diff
@@ -25,7 +25,58 @@ For the supervised training we use the following algorithms components:
...
+            --learning_parameters.early_stop_patience=8 \           # (for epoch runner) stop if the model don't improve within the number of epochs of patience
+            --learning_parameters.early_stop_iteration_patience=8 \ # (for iter runner) stop if the model don't improve within the number of iterations of patience
+
+    - `Supervised Contrastive Learning (SupCon) <https://arxiv.org/abs/2004.11362>`_: To enhance the performance of the algorithm in case when we have a small number of data.
+      More specifically, we train a model with two heads: segmentation head with Cross Entropy Loss and contrastive head with `DetCon loss <https://arxiv.org/abs/2103.10957>`_.
+      The below table shows how much performance (mDice) SupCon improved compared with baseline performance on the subsets of Pascal VOC 2012 with two classes (person, car) and three classes (person, car, bicycle).
```
Comment (kprokofi) [inline docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:42]: Maybe it is better to move it in the models section. Because we haven't described templates yet and we have dedicated table with performance in that section.

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst [hunk 4]
```diff
@@ -25,7 +25,58 @@ For the supervised training we use the following algorithms components:
...
+      The number of the subset datasets is 8, 16, and 24.
```
Comment (kprokofi) [inline docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:43]: the number of what? images? Can you clarify it here, please

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst [hunk 5]
```diff
@@ -25,7 +25,58 @@ For the supervised training we use the following algorithms components:
...
+      |                    | SL    | SupCon | SL    | SupCon | SL    | SupCon |
+      +--------------------+-------+--------+-------+--------+-------+--------+
+      | Lite-HRNet-s-mod2  | 60.36 | 60.86  | 63.75 | 66.84  | 66.43 | 69.81  |
+      +--------------------+-------+--------+-------+--------+-------+--------+
+      | Lite-HRNet-18-mod2 | 59.28 | 64.43  | 65.57 | 68.41  | 68.82 | 71.25  |
+      +--------------------+-------+--------+-------+--------+-------+--------+
+      | Lite-HRNet-x-mod3  | 59.98 | 63.36  | 63.60 | 68.63  | 68.33 | 71.21  |
+      +--------------------+-------+--------+-------+--------+-------+--------+
+
+      `three classes (person, car, bicycle)`
```
Comment (kprokofi) [inline docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:59]: I would suggest keeping only one table if it is okay, to not overwhelm with tables. For example, for the Semi-SL we also will choose only one label/unlabeled ratio. Is it possible to choose one table?

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst [hunk 6]
```diff
@@ -25,7 +25,58 @@ For the supervised training we use the following algorithms components:
...
+      |                    | SL    | SupCon | SL    | SupCon | SL    | SupCon |
+      +--------------------+-------+--------+-------+--------+-------+--------+
+      | Lite-HRNet-s-mod2  | 48.30 | 51.83  | 57.08 | 59.26  | 62.40 | 63.39  |
+      +--------------------+-------+--------+-------+--------+-------+--------+
+      | Lite-HRNet-18-mod2 | 53.47 | 54.90  | 56.69 | 60.32  | 62.81 | 64.56  |
+      +--------------------+-------+--------+-------+--------+-------+--------+
+      | Lite-HRNet-x-mod3  | 53.71 | 54.83  | 58.43 | 62.03  | 64.72 | 64.57  |
+      +--------------------+-------+--------+-------+--------+-------+--------+
+
+      You can use SupCon training like the below command.
```
Comment (kprokofi) [inline docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:73]: "The ``SupCon`` training can be launched by adding additional option to template parameters. It can be launched only with supervised (incremental) training type" (if it is true, of course, but my point is that we need to point out which training types are available for SupCon training)

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst [hunk 7]
```diff
@@ -109,7 +160,38 @@ For the `DIS5K <https://xuebinqin.github.io/dis/index.html>`_ we prepared random
...
-To be added soon
+Self-supervised learning can be one of the solutions if the user has a small data set, but label information is not yet available.
+General self-supervised Learning in academia is commonly used to obtain well-pretrained weights from a source dataset without label information.
+However, in real-world industries, it is difficult to apply because of small datasets, limited resources, or training in minutes.
+
+For these cases, OTX provides improved self-supervised learning recipes that can be applied to the above harsh environments.
+We adapted `DetCon <https://arxiv.org/abs/2103.10957>`_ as our self-supervised method.
+It takes some time to use these self-supervised learning recipes, but you can expect improved performance, especially in small-data regimes.
+
+The below table shows how much performance (mDice) self-supervised methods improved compared with baseline performance on the subsets of Pascal VOC 2012 with three classes (person, car, bicycle).
```
Comment (kprokofi) [inline docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:171]: If there is no labeled data for Self-SL, so what is the SL baseline? Can you clarify it here?

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst [hunk 8]
```diff
@@ -109,7 +160,38 @@ For the `DIS5K <https://xuebinqin.github.io/dis/index.html>`_ we prepared random
...
+The number of the subset datasets is 8, 16, and 24.
```
Comment (kprokofi) [inline docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:172]: Again, as for SupCon, number of what?

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst [hunk 9]
```diff
@@ -109,7 +160,38 @@ For the `DIS5K <https://xuebinqin.github.io/dis/index.html>`_ we prepared random
...
+|                    | SL    | Self-SL | SL    | Self-SL | SL    | Self-SL |
++--------------------+-------+---------+-------+---------+-------+---------+
+| Lite-HRNet-s-mod2  | 48.30 | 53.55   | 57.08 | 58.96   | 62.40 | 63.46   |
++--------------------+-------+---------+-------+---------+-------+---------+
+| Lite-HRNet-18-mod2 | 53.47 | 49.20   | 56.69 | 58.72   | 62.81 | 63.63   |
++--------------------+-------+---------+-------+---------+-------+---------+
+| Lite-HRNet-x-mod3  | 50.23 | 50.93   | 60.09 | 61.61   | 62.66 | 64.87   |
++--------------------+-------+---------+-------+---------+-------+---------+
+
+You can use Self-supervised learning like the below command.
```
Comment (kprokofi) [inline docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:188]: To enable Self-supervised training the command below can be executed:


Anyway, here we need some concrete example, bacause it is different training type. What should we pass to the data roots? For me, as from the user perspective - it is confusing. Please, describe more about the datasets, data roots and give a bit more concrete example how to launch this training type

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst [hunk 10]
```diff
@@ -109,7 +160,38 @@ For the `DIS5K <https://xuebinqin.github.io/dis/index.html>`_ we prepared random
 Self-supervised Learning
 ************************
 
-To be added soon
+Self-supervised learning can be one of the solutions if the user has a small data set, but label information is not yet available.
+General self-supervised Learning in academia is commonly used to obtain well-pretrained weights from a source dataset without label information.
+However, in real-world industries, it is difficult to apply because of small datasets, limited resources, or training in minutes.
+
+For these cases, OTX provides improved self-supervised learning recipes that can be applied to the above harsh environments.
```
Comment (kprokofi) [inline docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:167]: Please, describe more here about algo components. Are they the same as for supervised incremental learning or not? ( I mean: optimizer, augmentations, learning rate scheduler, etc. )

**Review by kprokofi — COMMENTED** [review]
Thank you for your work and great input! I added comments - we need to clarify some parts. Please, reflect these comment in another PR

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst (hunk 1 above)
Comment (sungchul1) [reply docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:28]: Done

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst (hunk 2 above)
Comment (sungchul1) [reply docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:36]: Done

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst (hunk 3 above)
Comment (sungchul1) [reply docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:42]: Done

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst (hunk 4 above)
Comment (sungchul1) [reply docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:43]: Done

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst (hunk 6 above)
Comment (sungchul1) [reply docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:73]: Done

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst (hunk 7 above)
Comment (sungchul1) [reply docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:171]: Done

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst (hunk 8 above)
Comment (sungchul1) [reply docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:172]: Done

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst (hunk 9 above)
Comment (sungchul1) [reply docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:188]: Done

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst (hunk 10 above)
Comment (sungchul1) [reply docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:167]: Self-SL has different components, but they are just for self-sl.
Self-sl training has two training stage (pretraining - fine tuning) and I added it.

File: docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst (hunk 5 above)
Comment (sungchul1) [reply docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:59]: Done
